### ISSUE TITLE
Support for static hints on std `replace()`  functions

### ISSUE THREAD

**Issue by Beinsezii** [microsoft/pyright#10354]
Python's `dataclasses` module offers a `replace()` which allows for a theoretically safe interface to update frozen structs

```python
import copy, dataclasses

@dataclasses.dataclass(frozen=True)
class Struc:
    a: int = 3
    b: str = "string"

data = Struc()
data2 = dataclasses.replace(data, a="word")  # Runs, but should lint regardless
data3 = copy.replace(data, c=False)  # Does not run, should lint
```

So while replace has a sig of `**changes: Any`, since we know in practice that it will `TypeError` if the wrong data is provided, to me it makes sense to have static checks for the args of `replace` against the field of the object provided, similar to standard assignment. The same could likely be applied to 3.13's `copy.replace`.

The reason I think this function in particular is suited to pyright unlike `setattr` or similar is that `replace` isn't intended for runtime reflection but rather a means for working with immutable classes.

**Comment by erictraut** [comment]
The function `dataclasses.replace` is defined in the typeshed stub `dataclasses.pyi`:

```python
def replace(obj: _DataclassT, /, **changes: Any) -> _DataclassT: ...
```

The function `copy.replace` is defined in the typeshed stub `copy.pyi`:

```python
def replace(obj: _SR, /, **changes: Any) -> _SR: ...
```

In both cases, pyright is correctly applying typing rules for these calls.

I think you're proposing for pyright to ignore the typeshed type definitions and override them with custom logic. That's technically possible, but it's something I try to avoid as a rule. I've broken this rule a few times in the past and I've come to regret it because it results in behaviors that are hard to explain. It also becomes a maintenance burden if and when the typeshed stubs change.

For these reasons, I'm going to reject this enhancement request. I'm willing to revisit this decision if there's sufficient signal from other pyright users in the form of upvotes. I'm going to close this issue for now because there are currently no plans to implement it.

**Comment by Beinsezii** [comment]
MyPy also just ignores it, and will report `dataclass.replace` being used with wrong keys/types.

**Comment by chazmo03** [comment]
Here's the MR for addressing this in mypy: https://github.com/python/mypy/pull/14849

**Comment by trim21** [comment]
I think this will be resolve by https://github.com/python/typeshed/pull/14819
